src/pgwire/src/protocol.rs

```diff
@@ -236,30 +236,44 @@ where
         let _ = session.vars_mut().set(&name, &value, local);
     }
 
-    // Register session with adapter.
-    let (mut adapter_client, startup) =
-        match adapter_client.startup(session, frontegg.is_some()).await {
-            Ok(startup) => startup,
-            Err(e) => {
-                return conn
-                    .send(ErrorResponse::from_adapter_error(Severity::Fatal, e))
-                    .await
-            }
-        };
-
-    let session = adapter_client.session();
     let mut buf = vec![BackendMessage::AuthenticationOk];
     for var in session.vars().notify_set() {
         buf.push(BackendMessage::ParameterStatus(var.name(), var.value()));
     }
     buf.push(BackendMessage::BackendKeyData {
         conn_id: session.conn_id(),
-        secret_key: startup.secret_key,
+        secret_key: session.secret_key(),
     });
+    // Immediately respond with connection success without waiting on the
+    // coordinator. This allows us to better meet SLA goals (like able to
+    // connect) at the expense of some specific problems:
+    // - Startup notices (like unknown database) won't be sent until the first
+    //   query.
+    // - An unknown username won't error until the first query. This won't be
+    //   noticed by users, though, since with frontegg enabled unknown users are
+    //   always created.
+    buf.push(BackendMessage::ReadyForQuery(session.transaction().into()));
+    conn.send_all(buf).await?;
+    conn.flush().await?;
+
+    // Register session with adapter.
+    let (adapter_client, startup) = match adapter_client.startup(session, frontegg.is_some()).await
+    {
+        Ok(startup) => startup,
+        Err(e) => {
+            return conn
+                .send(ErrorResponse::from_adapter_error(Severity::Fatal, e))
+                .await
```

Comment:
It does seem like it has the potential to be breaking the pgwire protocol because you end up with two responses for a single request. I'm familiar enough with the spec to know though.

Reply:
It won't be two responses, it'll just be a single unexpected error response then the connection closing.

Reply:
I guess if it works in `psql` then it's probably fine but it seems like we might be breaking spec by sending an unexpected response with no corresponding request.

Reply:
Yup! But since what follows is a disconnect this is probably ok. Successful user connections won't see anything out of spec.